pkg/api/validation/validation.go

```diff
@@ -1213,9 +1218,55 @@ func ValidateReplicationControllerSpec(spec *api.ReplicationControllerSpec) errs
 	} else {
 		labels := labels.Set(spec.Template.Labels)
 		if !selector.Matches(labels) {
-			allErrs = append(allErrs, errs.NewFieldInvalid("template.labels", spec.Template.Labels, "selector does not match template"))
+			allErrs = append(allErrs, errs.NewFieldInvalid("template.metadata.labels", spec.Template.Labels, "selector does not match template"))
+		}
+		allErrs = append(allErrs, ValidatePodTemplateSpec(spec.Template).Prefix("template")...)
+		if spec.Replicas > 1 {
+			allErrs = append(allErrs, ValidateReadOnlyPersistentDisks(spec.Template.Spec.Volumes).Prefix("template.spec.volumes")...)
 		}
-		allErrs = append(allErrs, ValidatePodTemplateSpec(spec.Template, spec.Replicas).Prefix("template")...)
+		// RestartPolicy has already been first-order validated as per ValidatePodTemplateSpec().
+		if spec.Template.Spec.RestartPolicy != api.RestartPolicyAlways {
+			allErrs = append(allErrs, errs.NewFieldValueNotSupported("template.spec.restartPolicy", spec.Template.Spec.RestartPolicy, []string{string(api.RestartPolicyAlways)}))
+		}
+	}
+	return allErrs
+}
+
+// ValidateDaemon tests if required fields in the daemon are set.
+func ValidateDaemon(controller *api.Daemon) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	allErrs = append(allErrs, ValidateObjectMeta(&controller.ObjectMeta, true, ValidateReplicationControllerName).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidateDaemonSpec(&controller.Spec).Prefix("spec")...)
+	return allErrs
+}
+
+// ValidateDaemonUpdate tests if required fields in the daemon are set.
+func ValidateDaemonUpdate(oldController, controller *api.Daemon) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	allErrs = append(allErrs, ValidateObjectMetaUpdate(&controller.ObjectMeta, &oldController.ObjectMeta).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidateDaemonSpec(&controller.Spec).Prefix("spec")...)
```

Comment:
Users can update object meta - the validation though should be only the
"create" validation (not the meta update validation) because this is a
template for a new item.

On Wed, Aug 5, 2015 at 3:19 PM, Ananya Kumar <notifications@github.com>
wrote:

> In pkg/api/validation/validation.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11801#discussion_r36342763>
> :
>
> > +	return allErrs
> > +}
> > +
> > +// ValidateDaemon tests if required fields in the daemon are set.
> > +func ValidateDaemon(controller *api.Daemon) errs.ValidationErrorList {
> > +	allErrs := errs.ValidationErrorList{}
> > +	allErrs = append(allErrs, ValidateObjectMeta(&controller.ObjectMeta, true, ValidateReplicationControllerName).Prefix("metadata")...)
> > +	allErrs = append(allErrs, ValidateDaemonSpec(&controller.Spec).Prefix("spec")...)
> > +	return allErrs
> > +}
> > +
> > +// ValidateDaemonUpdate tests if required fields in the daemon are set.
> > +func ValidateDaemonUpdate(oldController, controller *api.Daemon) errs.ValidationErrorList {
> > +	allErrs := errs.ValidationErrorList{}
> > +	allErrs = append(allErrs, ValidateObjectMetaUpdate(&controller.ObjectMeta, &oldController.ObjectMeta).Prefix("metadata")...)
> > +	allErrs = append(allErrs, ValidateDaemonSpec(&controller.Spec).Prefix("spec")...)
>
> Can I allow users to update ObjectMeta for the template? Right now we
> don't throw validation errors if users update ObjectMeta for an rc's pod
> template.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11801/files#r36342763>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift